macOS wxNativeFontInfo changes

csomor opened the pull request:
The next version of macOS will break our font info serialization of UI fonts, the PostScript names of which are regarded as private. Therefore we switch to the Apple recommended way of serializing the native font scriptors.

csomor committed:
adding native font descriptor serialization = v2

csomor committed:
remove common xml prefix from serialized string

vadz commented on src/osx/carbon/font.cpp:
```diff
@@ -934,8 +934,10 @@ CGFloat wxNativeFontInfo::GetCTSlant(CTFontDescriptorRef descr)
 }
 
+// common prefix of plist serializiation, gets removed and readded
+static const wxString s_plistPrefix = "<?xml version=\"1.0\" encoding=\"UTF-8\"?><!DOCTYPE plist PUBLIC "
+    "\"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">";
```
I think creating global `wxString`s does work right now, but we've had problems with this in the past, so I think it would be still better to replace it with an accessor function using static variable inside it, to avoid constructing it before wx is initialized.

csomor replied:
ok, done, thanks

vadz commented on src/osx/carbon/font.cpp:
```diff
@@ -949,86 +951,153 @@ bool wxNativeFontInfo::FromString(const wxString& s)
+        if (CFPropertyListIsValid(attributes, format))
+        {
+            wxCFDataRef listData(CFPropertyListCreateData(kCFAllocatorDefault, attributes, format, 0, NULL));
+            wxCFStringRef cfString( CFStringCreateFromExternalRepresentation( kCFAllocatorDefault, listData, kCFStringEncodingUTF8) );
```
Not directly related to this PR, but I wonder if it would be worth adding a `wxCFStringRef` ctor taking `wxCFDataRef`?

csomor replied:
I'll keep that in mind, right now this is the first occurence of CFStringCreateFromExternalRepresentation, so I'd leave this for the moment ...

vadz commented on src/osx/carbon/font.cpp:
```diff
@@ -949,86 +951,153 @@ bool wxNativeFontInfo::FromString(const wxString& s)
+            wxCFStringRef cfString( CFStringCreateFromExternalRepresentation( kCFAllocatorDefault, listData, kCFStringEncodingUTF8) );
+            wxString xml = cfString.AsString();
+            xml.Replace("\r",wxEmptyString,true);
+            if ( xml.StartsWith(s_plistPrefix ))
+                xml = xml.Mid(s_plistPrefix.length());
```
If we don't care about whether the XML prefix is present or not (although I wonder about it: shouldn't it always be present?), we could do this just as
```suggestion
            xml.StartsWith(s_plistPrefix, &xml);
```

vadz commented on src/osx/carbon/font.cpp:
```diff
@@ -949,86 +951,153 @@ bool wxNativeFontInfo::FromString(const wxString& s)
+            if ( xml.StartsWith(s_plistPrefix ))
+                xml = xml.Mid(s_plistPrefix.length());
+
+            s.Printf(wxT("%d;%d;%d;%d;"),
+                 2, // version
+                 GetUnderlined(),
+                 GetStrikethrough(),
+                 (int)GetEncoding());
+
+            s += xml;
```
Trivial, but as long we're using printf, why not construct the entire string using it, i.e.
```suggestion
            s.Printf("%d;%d;%d;%d;%s",
                 2, // version
                 GetUnderlined(),
                 GetStrikethrough(),
                 (int)GetEncoding(),
                 xml);
```

vadz reviewed (commented):
Thanks again for fixing this!

I'm just a bit concerned about the optionality of the "xml" PI prefix, IMO we need to either always have it or never do, I don't see any advantages in making it optional. Always stripping it would make the string shorter, so it seems slightly preferable, but always having it isn't really a problem neither. I'd just like it to be consistently one or the other.

csomor committed:
Update src/osx/carbon/font.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

csomor commented:
> I'm just a bit concerned about the optionality of the "xml" PI prefix, IMO we need to either always have it or never do, I don't see any advantages in making it optional. Always stripping it would make the string shorter, so it seems slightly preferable, but always having it isn't really a problem neither. I'd just like it to be consistently one or the other.

I'm only stripping it off, if it is precisely with the DTD of the propertylist 1.0, if it is something else (in the future?) it would be left there

csomor committed:
Update src/osx/carbon/font.cpp

Co-authored-by: VZ <vz-github@zeitlins.org>

csomor committed:
static string via accessor

vadz commented:
> I'm only stripping it off, if it is precisely with the DTD of the propertylist 1.0, if it is something else (in the future?) it would be left there

Hmm, wouldn't we need to change the version anyhow if the XML format changes in the future?

csomor commented:
> > I'm only stripping it off, if it is precisely with the DTD of the propertylist 1.0, if it is something else (in the future?) it would be left there
> 
> Hmm, wouldn't we need to change the version anyhow if the XML format changes in the future?

I didn't think of that, but yes, we could increase it if the plist format is not 1.0 anymore .... ok, so I'll do that unconditionally

csomor committed:
striping off xml preamble unconditionally

if we use a different format in the future, we will have to increase our version number

csomor committed:
applying italic directly to the font descriptor

thus preserving attributes in the native font descriptor which we don’t store explicitly yet

csomor committed:
Adding support for preserving font width

Although we don’t express this in the public API yet, we try to preserve eg condensed, when changing the font width

vslavik commented on src/osx/carbon/font.cpp:
```diff
@@ -949,86 +969,151 @@ bool wxNativeFontInfo::FromString(const wxString& s)
+        wxFontEncoding encoding = (wxFontEncoding)l;
+
+        wxString xml = tokenizer.GetString();
+        xml+=GetPListPrefix();
```
Shouldn't this be `xml = GetPlistPrefix() + xml;` ?

csomor replied:
of course, thanks a lot, how embarrassing ....

vslavik commented on src/osx/carbon/font.cpp:
```diff
@@ -949,86 +969,151 @@ bool wxNativeFontInfo::FromString(const wxString& s)
+        }
+    }
+
+    if ( s.empty() )
+    {
+        // fallback to version 1
```
Do you know if/when this can happen? `CTFontDescriptorCopyAttributes()` looks like it doesn't return NULL per docs. Nor is it clear to me how could the dictionary fail the `CFPropertyListIsValid()` test?

My concern here is that I think the serialization format revision should be bumped up to 2, with v1 remaining only reading-only backward compatibility code (which can be removed down the road when *N* years pass) instead of continuing to serialize in both v1 and v2 formats simultaneously.

csomor replied:
I've copied that code from the Apple Font docs in the serialization part

https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/CoreText_Programming/FontOperations/FontOperations.html#//apple_ref/doc/uid/TP40005533-CH4-SW25

so I thought if they assume things could go wrong, I should do so too, and fall back in case of errors to what we know, so in practice this should not happen, but we'd be safe

vslavik reviewed (commented):
Thanks for doing this! No other comments except the two inline ones.

csomor committed:
Adding reference, bug fix

double checked the font weight constants,
